### PR TITLE
Fix error handling in asn1_item_embed_new.

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Unfortunately I did not completely fix the error handling in asn1_item_embed_new.

After I managed to repeat my memory error injection tests with OpenSSL 1.1.0
the following crash happened faily often:

```
==3547==ERROR: AddressSanitizer: heap-use-after-free on address 0x613000b9fdc0 at pc 0x00000213712a bp 0x7ff745698410 sp 0x7ff745698408
READ of size 8 at 0x613000b9fdc0 thread T146
    #0 0x2137129 in asn1_do_lock crypto/asn1/tasn_utl.c:79
    #1 0x2132886 in asn1_item_embed_free crypto/asn1/tasn_fre.c:92
    #2 0x2132886 in ASN1_item_ex_free crypto/asn1/tasn_fre.c:28
    #3 0x2136527 in asn1_item_embed_new crypto/asn1/tasn_new.c:145
    #4 0x2136527 in ASN1_item_ex_new crypto/asn1/tasn_new.c:39
    #5 0x2128ba3 in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:290
    #6 0x2128ba3 in ASN1_item_ex_d2i crypto/asn1/tasn_dec.c:114
    #7 0x2128ba3 in ASN1_item_d2i crypto/asn1/tasn_dec.c:104
    #8 0x22e13d2 in ASN1_item_d2i_bio crypto/asn1/a_d2i_fp.c:70
    #9 0x226e5f9 in X509_load_cert_file crypto/x509/by_file.c:113
    #10 0x226e5f9 in by_file_ctrl crypto/x509/by_file.c:68
    #11 0x1dbe514 in OpcUa_P_OpenSSL_PKI_OpenCertificateStore platforms/linux/opcua_p_openssl_pki.c:203
    #12 0x1d44f57 in OpcUa_SslSocket_VerifyCertificate platforms/linux/opcua_p_socket_ssl.c:524
    #13 0x2086598 in ssl_verify_cert_chain ssl/ssl_cert.c:437
    #14 0x20d7e4f in tls_process_client_certificate ssl/statem/statem_srvr.c:2892
    #15 0x20ab1d3 in read_state_machine ssl/statem/statem.c:591
    #16 0x20ab1d3 in state_machine ssl/statem/statem.c:387
    #17 0x2070c16 in ssl3_read_bytes ssl/record/rec_layer_s3.c:1031
    #18 0x20816c7 in ssl3_read_internal ssl/s3_lib.c:3838
    #19 0x20816c7 in ssl3_read ssl/s3_lib.c:3862
    #20 0x208f08f in SSL_read ssl/ssl_lib.c:1552
    #21 0x1d456a5 in OpcUa_P_SocketService_SslRead platforms/linux/opcua_p_socket_ssl.c:248
    #22 0x1db5373 in OpcUa_HttpsStream_Receive transport/https/opcua_httpsstream.c:2118
    #23 0x1db5373 in OpcUa_HttpsStream_DataReady transport/https/opcua_httpsstream.c:2664
    #24 0x1da6723 in OpcUa_HttpsListener_ReadEventHandler transport/https/opcua_httpslistener.c:1251
    #25 0x1da0bd5 in OpcUa_HttpsListener_EventCallback transport/https/opcua_httpslistener.c:1664
    #26 0x1d45bc3 in OpcUa_SslSocket_DoStateMachine platforms/linux/opcua_p_socket_ssl.c:171
    #27 0x1d46432 in OpcUa_SslSocket_InternalRead platforms/linux/opcua_p_socket_ssl.c:782
    #28 0x1d46432 in OpcUa_RawSocket_EventCallback platforms/linux/opcua_p_socket_ssl.c:839
    #29 0x1d3d313 in OpcUa_Socket_HandleEvent platforms/linux/opcua_p_socket_internal.c:898
    #30 0x1d3f644 in OpcUa_P_Socket_HandleFdSet platforms/linux/opcua_p_socket_internal.c:1336
    #31 0x1d420ba in OpcUa_P_SocketManager_ServeLoopInternal platforms/linux/opcua_p_socket_internal.c:1123
    #32 0x1d420ba in OpcUa_SocketManager_AcceptHandlerThread platforms/linux/opcua_p_socket_internal.c:723
    #33 0x1d4910d in pthread_start platforms/linux/opcua_p_thread.c:46
    #34 0x7ff77040be99 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x7e99)
    #35 0x7ff76f49b2ec in clone (/lib/x86_64-linux-gnu/libc.so.6+0xf62ec)

0x613000b9fdc0 is located 320 bytes inside of 328-byte region [0x613000b9fc80,0x613000b9fdc8)
freed by thread T146 here:
    #0 0x7ff7706fa4f8 in __interceptor_free ../../../../gcc-8-20170611/libsanitizer/asan/asan_malloc_linux.cc:45
    #1 0x213320c in asn1_item_embed_free crypto/asn1/tasn_fre.c:119
    #2 0x213320c in ASN1_item_ex_free crypto/asn1/tasn_fre.c:28
    #3 0x2135684 in asn1_item_embed_new crypto/asn1/tasn_new.c:145
    #4 0x2135684 in asn1_template_new crypto/asn1/tasn_new.c:235
    #5 0x2136518 in asn1_item_embed_new crypto/asn1/tasn_new.c:132
    #6 0x2136518 in ASN1_item_ex_new crypto/asn1/tasn_new.c:39
    #7 0x2128ba3 in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:290
    #8 0x2128ba3 in ASN1_item_ex_d2i crypto/asn1/tasn_dec.c:114
    #9 0x2128ba3 in ASN1_item_d2i crypto/asn1/tasn_dec.c:104
    #10 0x22e13d2 in ASN1_item_d2i_bio crypto/asn1/a_d2i_fp.c:70
    #11 0x226e5f9 in X509_load_cert_file crypto/x509/by_file.c:113
    #12 0x226e5f9 in by_file_ctrl crypto/x509/by_file.c:68
    #13 0x1dbe514 in OpcUa_P_OpenSSL_PKI_OpenCertificateStore platforms/linux/opcua_p_openssl_pki.c:203
    #14 0x1d44f57 in OpcUa_SslSocket_VerifyCertificate platforms/linux/opcua_p_socket_ssl.c:524
    #15 0x2086598 in ssl_verify_cert_chain ssl/ssl_cert.c:437
    #16 0x20d7e4f in tls_process_client_certificate ssl/statem/statem_srvr.c:2892
    #17 0x20ab1d3 in read_state_machine ssl/statem/statem.c:591
    #18 0x20ab1d3 in state_machine ssl/statem/statem.c:387
    #19 0x2070c16 in ssl3_read_bytes ssl/record/rec_layer_s3.c:1031
    #20 0x20816c7 in ssl3_read_internal ssl/s3_lib.c:3838
    #21 0x20816c7 in ssl3_read ssl/s3_lib.c:3862
    #22 0x208f08f in SSL_read ssl/ssl_lib.c:1552
    #23 0x1d456a5 in OpcUa_P_SocketService_SslRead platforms/linux/opcua_p_socket_ssl.c:248
    #24 0x1db5373 in OpcUa_HttpsStream_Receive transport/https/opcua_httpsstream.c:2118
    #25 0x1db5373 in OpcUa_HttpsStream_DataReady transport/https/opcua_httpsstream.c:2664
    #26 0x1da6723 in OpcUa_HttpsListener_ReadEventHandler transport/https/opcua_httpslistener.c:1251
    #27 0x1da0bd5 in OpcUa_HttpsListener_EventCallback transport/https/opcua_httpslistener.c:1664
    #28 0x1d45bc3 in OpcUa_SslSocket_DoStateMachine platforms/linux/opcua_p_socket_ssl.c:171
    #29 0x1d46432 in OpcUa_SslSocket_InternalRead platforms/linux/opcua_p_socket_ssl.c:782
    #30 0x1d46432 in OpcUa_RawSocket_EventCallback platforms/linux/opcua_p_socket_ssl.c:839
    #31 0x1d3d313 in OpcUa_Socket_HandleEvent platforms/linux/opcua_p_socket_internal.c:898
    #32 0x1d3f644 in OpcUa_P_Socket_HandleFdSet platforms/linux/opcua_p_socket_internal.c:1336
    #33 0x1d420ba in OpcUa_P_SocketManager_ServeLoopInternal platforms/linux/opcua_p_socket_internal.c:1123
    #34 0x1d420ba in OpcUa_SocketManager_AcceptHandlerThread platforms/linux/opcua_p_socket_internal.c:723
    #35 0x1d4910d in pthread_start platforms/linux/opcua_p_thread.c:46
    #36 0x7ff77040be99 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x7e99)

previously allocated by thread T146 here:
    #0 0x7ff7706fa850 in __interceptor_malloc ../../../../gcc-8-20170611/libsanitizer/asan/asan_malloc_linux.cc:62
    #1 0x21dd4f1 in CRYPTO_malloc crypto/mem.c:76
    #2 0x21dd4f1 in CRYPTO_zalloc crypto/mem.c:100
    #3 0x213644c in asn1_item_embed_new crypto/asn1/tasn_new.c:122
    #4 0x213644c in ASN1_item_ex_new crypto/asn1/tasn_new.c:39
    #5 0x2128ba3 in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:290
    #6 0x2128ba3 in ASN1_item_ex_d2i crypto/asn1/tasn_dec.c:114
    #7 0x2128ba3 in ASN1_item_d2i crypto/asn1/tasn_dec.c:104
    #8 0x22e13d2 in ASN1_item_d2i_bio crypto/asn1/a_d2i_fp.c:70
    #9 0x226e5f9 in X509_load_cert_file crypto/x509/by_file.c:113
    #10 0x226e5f9 in by_file_ctrl crypto/x509/by_file.c:68
    #11 0x1dbe514 in OpcUa_P_OpenSSL_PKI_OpenCertificateStore platforms/linux/opcua_p_openssl_pki.c:203
    #12 0x1d44f57 in OpcUa_SslSocket_VerifyCertificate platforms/linux/opcua_p_socket_ssl.c:524
    #13 0x2086598 in ssl_verify_cert_chain ssl/ssl_cert.c:437
    #14 0x20d7e4f in tls_process_client_certificate ssl/statem/statem_srvr.c:2892
    #15 0x20ab1d3 in read_state_machine ssl/statem/statem.c:591
    #16 0x20ab1d3 in state_machine ssl/statem/statem.c:387
    #17 0x2070c16 in ssl3_read_bytes ssl/record/rec_layer_s3.c:1031
    #18 0x20816c7 in ssl3_read_internal ssl/s3_lib.c:3838
    #19 0x20816c7 in ssl3_read ssl/s3_lib.c:3862
    #20 0x208f08f in SSL_read ssl/ssl_lib.c:1552
    #21 0x1d456a5 in OpcUa_P_SocketService_SslRead platforms/linux/opcua_p_socket_ssl.c:248
    #22 0x1db5373 in OpcUa_HttpsStream_Receive transport/https/opcua_httpsstream.c:2118
    #23 0x1db5373 in OpcUa_HttpsStream_DataReady transport/https/opcua_httpsstream.c:2664
    #24 0x1da6723 in OpcUa_HttpsListener_ReadEventHandler transport/https/opcua_httpslistener.c:1251
    #25 0x1da0bd5 in OpcUa_HttpsListener_EventCallback transport/https/opcua_httpslistener.c:1664
    #26 0x1d45bc3 in OpcUa_SslSocket_DoStateMachine platforms/linux/opcua_p_socket_ssl.c:171
    #27 0x1d46432 in OpcUa_SslSocket_InternalRead platforms/linux/opcua_p_socket_ssl.c:782
    #28 0x1d46432 in OpcUa_RawSocket_EventCallback platforms/linux/opcua_p_socket_ssl.c:839
    #29 0x1d3d313 in OpcUa_Socket_HandleEvent platforms/linux/opcua_p_socket_internal.c:898
    #30 0x1d3f644 in OpcUa_P_Socket_HandleFdSet platforms/linux/opcua_p_socket_internal.c:1336
    #31 0x1d420ba in OpcUa_P_SocketManager_ServeLoopInternal platforms/linux/opcua_p_socket_internal.c:1123
    #32 0x1d420ba in OpcUa_SocketManager_AcceptHandlerThread platforms/linux/opcua_p_socket_internal.c:723
    #33 0x1d4910d in pthread_start platforms/linux/opcua_p_thread.c:46
    #34 0x7ff77040be99 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x7e99)
```

I think the main difference to 1.0.2 is the branch from asn1_do_lock to the error handling.
This is what likely happened in the above call stack.  Previously asn1_do_lock was not done
for embedded objects, and it's retrun value needed no checking.

Not sure if that needs backporting to 1.0.2 though, apparently it might depend on the
data structures, of an error in the embedded objects can happen.

### CODE REVIEW

File: crypto/asn1/tasn_new.c [hunk 1]
```diff
@@ -41,6 +41,7 @@ int ASN1_item_ex_new(ASN1_VALUE **pval, const ASN1_ITEM *it)
 
 int asn1_item_embed_new(ASN1_VALUE **pval, const ASN1_ITEM *it, int embed)
 {
+    int err = ERR_R_MALLOC_FAILURE;
```
Comment (FdaSilvaYY) [inline crypto/asn1/tasn_new.c]: There is a `int `declared 6 line below

File: crypto/asn1/tasn_new.c [hunk 2]
```diff
@@ -142,22 +143,34 @@ int asn1_item_embed_new(ASN1_VALUE **pval, const ASN1_ITEM *it, int embed)
...
-    ASN1err(ASN1_F_ASN1_ITEM_EMBED_NEW, ERR_R_MALLOC_FAILURE);
+    ASN1err(ASN1_F_ASN1_ITEM_EMBED_NEW, err);
 #ifndef OPENSSL_NO_CRYPTO_MDEBUG
     OPENSSL_mem_debug_pop();
 #endif
     return 0;
 
  auxerr2:
-    ASN1_item_ex_free(pval, it);
+    err = ASN1_R_AUX_ERROR;
```
Comment (FdaSilvaYY) [inline crypto/asn1/tasn_new.c]: These `auxerr1` labels  can be moved above `memerr2 `too ?

File: crypto/asn1/tasn_new.c (hunk 1 above)
Comment (bernd-edlinger) [inline crypto/asn1/tasn_new.c]: Yes,

would you prefer

 ```
 int err = ERR_R_MALLOC_FAILURE;
 int i;
```
or

`int i, err = ERR_R_MALLOC_FAILURE;
`

**Comment by FdaSilvaYY** [comment]
Second is ok for me.

**Comment by bernd-edlinger** [comment]
Okay see updated patch, how do you like it now?

File: crypto/asn1/tasn_new.c (hunk 2 above)
Comment (bernd-edlinger) [inline crypto/asn1/tasn_new.c]: Yes, I am open for suggestions to find the least-ugly solution here.
Following @FdaSilvaYY 's comment I already lined up the different
labels avoiding the back-jumps.

**Comment by bernd-edlinger** [comment]
BTW: this patch works for master & 1.1.0
1.0.2 will probably need a different PR.

**Comment by richsalz** [comment]
I could not think of anything better than what you have right now, except perhaps to set the error code *before* doing the jumps, which might reduce the need for the number of labels.  I don’t know if that will be better or not.

File: crypto/asn1/tasn_new.c [hunk 3]
```diff
@@ -46,7 +46,7 @@ int asn1_item_embed_new(ASN1_VALUE **pval, const ASN1_ITEM *it, int embed)
     const ASN1_AUX *aux = it->funcs;
     ASN1_aux_cb *asn1_cb;
     ASN1_VALUE **pseqval;
-    int i;
+    int i, err = ERR_R_MALLOC_FAILURE;
```
Comment (FdaSilvaYY) [inline crypto/asn1/tasn_new.c]: blank line after decl ;)

**Comment by FdaSilvaYY** [comment]
There is just 4 'goto memerr#', if I'm right.

File: crypto/asn1/tasn_new.c (hunk 3 above)
Comment (bernd-edlinger) [inline crypto/asn1/tasn_new.c]: Yes, thanks :)

File: crypto/asn1/tasn_new.c [hunk 4]
```diff
@@ -141,19 +141,27 @@ int asn1_item_embed_new(ASN1_VALUE **pval, const ASN1_ITEM *it, int embed)
...
+    goto memerr1;
+ auxerr2:
+    err = ASN1_R_AUX_ERROR;
  memerr2:
-    ASN1_item_ex_free(pval, it);
+    for (tt = it->templates; i > 0; i--, tt++) {
+        pseqval = asn1_get_field_ptr(pval, tt);
+        asn1_template_free(pseqval, tt);
+    }
+    asn1_do_lock(pval, -1, it);
```
Comment (bernd-edlinger) [inline crypto/asn1/tasn_new.c]: There is also an "asn1_enc_free" which is formally the inverse of "asn_enc_init".
I should not make a difference in practice at least here.  Should I add that call here?

**Comment by bernd-edlinger** [comment]
I did add an empty line after the variable section.

**Comment by bernd-edlinger** [comment]
@levitte last time I touched this function you did the review, so I think you are the right expert
for a second review here.
Thanks.

File: crypto/asn1/tasn_new.c (hunk 4 above)
Comment (levitte) [inline crypto/asn1/tasn_new.c]: This part here seems over the top.  I would need some *very* strong arguments to find code copying acceptable, so unless you have them, I'd think that you should put back the call to `ASN1_item_ex_free`...

**Comment by levitte** [comment]
Frankly, this change has me going veeeeery nervous.  Details get forgotten because there are issues with the standard freeing system.  Wouldn't the original issue just as well be fixed by checking that `*lock != NULL` in the right spot in `asn1_do_lock`?

**Comment by bernd-edlinger** [comment]
This is something that did not happen with 1.0.2 but probably just by chance.

In the call stack above, the place where the memory was free'd looks like
asn1_item_embed_new->asn1_template_new->asn1_item_embed_new->ASN1_item_ex_free,
but the place where the memory was used looks like asn1_item_embed_new->ASN1_item_ex_free.

I believe the the problem is that asn1_item_embed_new (embed=1) free'd the original object,
and after asn1_template_new returns with error the object is only partially constructed,
and it depends on the embed parameter if the underlying storage has to be free'd.

So this PR rolls the template construction back step by step, and that completely
different from what ASN1_item_ex_free does or even asn1_item_embed_free,
which I considered making non-static, bul looking closely it is just something
different.

**Comment by bernd-edlinger** [comment]
... and AFAICT the issue is not that `*lock` is `NULL` but that `lock` points to free'd memory.

**Comment by bernd-edlinger** [comment]
An alternative with ASN1_item_ex_free would look like this:
```
 auxerr:
    err = ASN1_R_AUX_ERROR;
    goto memerr;
 auxerr2:
    err = ASN1_R_AUX_ERROR;
 memerr2:
    if (embed == 0) {
        ASN1_item_ex_free(pval, it);
        OPENSSL_free(*pval);
        *pval = NULL;
    }
 memerr:
    ASN1err(ASN1_F_ASN1_ITEM_EMBED_NEW, err);
#ifndef OPENSSL_NO_CRYPTO_MDEBUG
    OPENSSL_mem_debug_pop();
#endif
    return 0;
```

BUT the object is only partially constructed at this point, therefore
ASN1_item_ex_free may fail internally and ASN1_OP_FREE_PRE
may be called without ASN1_OP_NEW_PRE ever being called.

**Comment by bernd-edlinger** [comment]
I still think this is the correct way to fix the error-handling here.
@levitte the only thing that makes me nervous about it, is how easy
 it was to reproduce in 1.1.0/master. So do you still uphold your objections?

**Comment by bernd-edlinger** [comment]
Ping.
Any ideas how to get this fixed?

**Comment by bernd-edlinger** [comment]
Ping?

**Comment by richsalz** [comment]
ping @levitte -- are your concerns addressed?

**Comment by bernd-edlinger** [comment]
Ping?

**Comment by mattcaswell** [comment]
@levitte may not be around much this week

**Pull request closed without merging**